keys.go

```diff
@@ -0,0 +1,41 @@
+package main
+
+import (
+	"crypto/ed25519"
+	crand "crypto/rand"
+	"crypto/sha256"
+	"io"
+
+	libp2p "github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/mr-tron/base58"
+	"golang.org/x/crypto/hkdf"
+)
+
+const seedBytes = 32
+
+// newSeed returns a b58 encoded random seed.
+func newSeed() (string, error) {
+	bs := make([]byte, seedBytes)
+	_, err := io.ReadFull(crand.Reader, bs)
+	if err != nil {
+		return "", err
+	}
+	return base58.Encode(bs), nil
+}
+
+// derive derives libp2p keys from a b58-encoded seed.
+func deriveKey(b58secret string, info []byte) (libp2p.PrivKey, error) {
+	secret, err := base58.Decode(b58secret)
```

Comment:
What if the secret is too short? It's important to have a cryptographically secure secret here and not just a random passphrase the user remembers.